### ISSUE TITLE
Support for Multi-Tenancy

### ISSUE THREAD

**Issue by altruistcoder** [mlflow/mlflow#5844]
Hello All,

I have an MLFlow Instance deployed on an Openshift cluster where the backend store is running as a PostGreSQL Instance and the Artifact Store is taken as a bucket in our Rook Ceph cluster..

I have two questions to ask:

1. First, is it possible to add Multi-tenancy to a single instance of MLFlow so as to filter Experiments, Backend Store and Artifacts Store on the basis of User Authentication.
2. And, if yes, can you please point me to any documentation, instructions or manuals from where I can get more info on how can I achieve this?

**Comment by nreith** [comment]
The approach I am considering is possibly to put a wrapper or decorator around most mlfliw api calls to init and switch db per team based on inputs. Others have already done basic auth with nginx around mlflow. If you use s3 as the artifact store and don't proxy artifacts through the tracking server, they upload there and at the same time that bucket name (i.e. username) and token are the credentials. The same auth is used for nginx and the bucket name can also be team name and db schema name. Then the wrapper or decorator in would either reinitialize the db and async wait to point to the right place before querying. Or it could reinit the whole flask app/server if easier with the new info.

That doesn't get to the ui yet. But I assume a similar approach could be taken. Also the ui can be a bit less finicky because it can run anywhere on demand and doesn't need to be up or accessible 24/7. Others have put it inside of jupyterhub. It could even just be run locally.

**Comment by umialpha** [comment]
our company needs this feature so much

**Comment by nreith** [comment]
I suspect this is part of the point. The core app has been open sourced including v 2.x recently. But the enterprise parts are part of Databricks' product.

**Comment by ketangangal** [comment]
Looking for this feature to be implemented as open source as well.

**Comment by amansharif** [comment]
Has there been any update on this feature ? It's been long since there's any activity so wondering if we should expect this feature as open source or not.

**Comment by maistrotoad** [comment]
➕

**Comment by bobbravo2** [comment]
➕

**Comment by nreith** [comment]
My experience with most "enterprise" open-source is that they give the core features free, but hold out on the enterprise bits like RBAC, multi-tenant and SSO and want companies to pay for those. Having done a deep dive on the code, you can see shadows on the wall of the cave Plato style of how Databricks does have these features in their codebase elsewhere, by the naming and code snippets in the free version of MLFLow, but you would need to be really good and lucky to guess and reverse engineer all of that with your own code to integrate with those bits.

What we ended up doing and I find to be an easier and more sustainable pattern is to just give each user or group of users (ad group), their own server or multiple mlflow servers they can share, each of which is a standalone container running an isolated mlflow instance. Then you just need some sort of container lifecycle management/orchestration tool to make it easier than `kubectl apply deployment.yaml` or Helm equivalent for users to click and get a new instance. It took quite a bit of work, but JupyterHub is a nice framework for this. Kubeflow has namespace/user/team isolation, and I believe already integrates with MLFlow nicely. Kubero is another that is Heroku-like, and has lots of supported open-source tools you can spin up with a single click.

tldr; I found it was too much code surgery to shoehorn these features into this MLFlow codebase, and then would need to be repeated for every single open-source tool with a similar "freemium" model.

**Comment by mprahl** [comment]
Hello! I’m a Kubeflow Pipelines maintainer exploring a first-class integration of MLflow into the Kubeflow ecosystem. I believe users would benefit from combining MLflow’s world‑class experiment tracking and model management with Kubeflow’s AI/ML Kubernetes native ecosystem.

Kubeflow's primary deployment pattern is to scope tenancy to Kubernetes namespaces. Would the MLflow community be open to an optional, backward-compatible multi-tenancy mode aligned to namespaces/workspaces? These aren't Kubernetes namespaces, but just a generic concept that we could adapt to Kubernetes namespaces.

This is an initial draft to start the discussion. I’d love your feedback and I'm happy to adapt the proposal based on community input.

### Kubeflow's Motivation

- Enable secure, scalable multi-tenant deployments on Kubernetes without separate MLflow instances.
- Align with existing patterns in Kubeflow and cluster RBAC.

### Proposal (opt-in)

- **Data model**
  - Add a nullable, platform‑agnostic `namespace` column to top‑level resources (e.g., experiments, registered models, prompts). This is a generic grouping key (e.g., tenant/team/project), not inherently tied to Kubernetes.
  - Child resources inherit their parent’s namespace (e.g., runs inherit experiment namespace).
  - When multi‑tenancy is disabled, `namespace` remains NULL and behavior is unchanged.

- **Client SDK, CLI, and REST**
  - Add `mlflow.set_namespace("team-a")` and an `MLFLOW_NAMESPACE` env var (similar to tracking URI configurations).
  - Clients send `X-MLflow-Namespace` on all calls when set.
  - All MLflow UI URLs are logged with the `?namespace=` query argument for the UI to select the correct namespace.

- **Server behavior**
  - When multi‑tenancy mode is enabled (via server configuration), list/search APIs are filtered by the `X-MLflow-Namespace` value; resource creation assigns that namespace.
  - Cross‑namespace API calls would not be allowed (e.g. I can't search for experiments across multiple namespaces).
  - If the header is missing while multi‑tenancy is enabled, return an error.
 
- **Platform Specific Implementations**
  - Introduce a pluggable `NamespaceProvider` to enumerate namespaces available to a user. A Kubeflow implementation would list Kubernetes namespaces/Kubeflow Profiles, but providers could map to any tenancy construct (e.g., projects, teams, orgs).
  - Expose Flask request context variables (e.g., `g.mlflow_namespace`, `g.mlflow_rbac_resource`, `g.mlflow_resource_id`) so auth plugins can enforce RBAC. In Kubeflow, we'd use Kubernetes SubjectAccessReview for authorize/read/write within a namespace.

- **UI**
  - Optional namespace selector, with results filtered by the selected namespace.
  - All API requests set the `X-MLflow-Namespace` header.
  - The UI would keep a `?namespace=` query argument in the URL to allow easy URL sharing between users.

### Backward compatibility and migration
- Feature is disabled by default; no behavior change.
- Schema migrations add nullable `namespace` columns; existing rows remain NULL.
- If admins switch to multitenancy mode, we can treat empty/null namespaces as a special reserved namespace of "default".

If there's interest, I can draft a short design doc and contribute a reference plugin integrating with Kubernetes namespaces and SubjectAccessReview. Thank you!

@harupy and @BenWilson2 , I'm tagging you since I've seen you've both been recently active MLflow maintainers. Please include any others to the conversation. 😄

**Comment by B-Step62** [comment]
Thank you for the proposal @mprahl! With the amount of request and community interests, we are definitely interested in moving this idea forward.

As for the solution, backward compatibility and minimizing impact to existing workload is the top priority, so it is great your solution already accounts for it! The devil is in the details, we need to think carefully and look through the end-to-end platform. We discussed among maintainers and came up with a few questions and suggestions, but I think it would be more productive if we move the discussion on a document, rather than single-track comments here. Would you mind drafting a RFC with your proposal (and alternatives if you have)?

Template: https://docs.google.com/document/d/1AQGgJk-hTkUo0lTkGqCGQOMelQmz05kQz_OA4bJWaJE/edit?tab=t.0#heading=h.4cz970y1mk93

**Comment by mprahl** [comment]
> Thank you for the proposal [@mprahl](https://github.com/mprahl)! With the amount of request and community interests, we are definitely interested in moving this idea forward.
> 
> As for the solution, backward compatibility and minimizing impact to existing workload is the top priority, so it is great your solution already accounts for it! The devil is in the details, we need to think carefully and look through the end-to-end platform. We discussed among maintainers and came up with a few questions and suggestions, but I think it would be more productive if we move the discussion on a document, rather than single-track comments here. Would you mind drafting a RFC with your proposal (and alternatives if you have)?
> 
> Template: https://docs.google.com/document/d/1AQGgJk-hTkUo0lTkGqCGQOMelQmz05kQz_OA4bJWaJE/edit?tab=t.0#heading=h.4cz970y1mk93

@B-Step62 thanks for getting back to me! Here is the proposal, and I apologize for the length but I wanted to be thorough: 😄 
https://docs.google.com/document/d/1IbFfceCmWV3knJfc0ninn58EXLVbHUh1meV_pknOM40/edit

**Comment by juliusvonkohout** [comment]
Lets please opt for a proper solution and not "Separate MLflow instances per tenant" so "Workspace-aware MLflow server"

**Comment by ask664** [comment]
Hi @mprahl 
I’m interested in contributing to the Multi-Tenancy (Workspaces-aware) support that’s currently under design for MLflow. Our organization is also implementing a similar capability internally, and my team would be keen to collaborate and contribute to this initiative.
Could you please include me in any upcoming community calls or discussions related to this feature? I’d appreciate the opportunity to better understand the proposal, roadmap, and potential contribution areas.

Adding to @B-Step62  for some guidance on how to plug in and get involved.

Thank you!

**Comment by B-Step62** [comment]
@mprahl Fantastic write out, thank you so much for the very thoughtful and comprehensive design doc!

Did a first scan and left some questions, but the overall server-side enforcement approach makes much sense. I will let other maintainers take a look during their day. Generally the contents are solid so I'm hoping to close the loop soon and get to implementation phase quick.

@ask664 Thank you for the offering, we will keep the discussion transparent here and on the design doc. Once the design is finalized and the task plan is sorted out, we can start calling for contributions.

**Comment by mprahl** [comment]
> [@mprahl](https://github.com/mprahl) Fantastic write out, thank you so much for the very thoughtful and comprehensive design doc!
> 
> Did a first scan and left some questions, but the overall server-side enforcement approach makes much sense. I will let other maintainers take a look during their day. Generally the contents are solid so I'm hoping to close the loop soon and get to implementation phase quick.

Thanks so much for the review @B-Step62! I responded/addressed your comments.

I'm actively working on a proof of concept leveraging this design, so I'll share that soon!

**Comment by nsh87** [comment]
Please do this yesterday 😀

Considering forking this and writing/reading a user OAuth identity token on every table row to get some basic user-isolation. Not ideal.

**Comment by mprahl** [comment]
> Please do this yesterday 😀
> 
> Considering forking this and writing/reading a user OAuth identity token on every table row to get some basic user-isolation. Not ideal.

I should have a PR up in the next few days adding the backend, auth, and SDK capabilities based on the design at https://docs.google.com/document/d/1IbFfceCmWV3knJfc0ninn58EXLVbHUh1meV_pknOM40/edit.

Once/if the backend PR gets accepted, I'll work on the UI changes which are fairly minimal (already works locally).

So hold off on any forks and please provide feedback once the PR is up 😆

**Comment by mprahl** [comment]
@BenWilson2 @B-Step62 @ask664 thank you for all of your reviews on the design document! I've submitted a PR for the backend, authorization, SDK, and docs at #18869.

Unfortunately, it's a massive PR due to the docs and tests. Please let me know how I can help with the review process. @Gkrumbach07, a fellow Kubeflow community member, will be posting a separate PR with the UI support.

**Comment by andreas-repp** [comment]
Hey everyone!

Thanks for the amazing work @mprahl and the maintainers supporting you with your PRs! We were missing this feature so much and are currently looking into setting up Kubeflow and this is a major pain point for us. 

Without intending any pressure and without any expectations of correctness: Might it be possible to get a guesstimate when this feature might be complete in MLFlow? And @mprahl: Do you have an idea when it could then be available / fully utilized in Kubeflow? Maybe I missed it in the thread, but is there a related Kubeflow ticket? 

The reason I'm asking is that we are currently in the process of setting up Kubeflow with MLFlow and based on the quick progress on this over the last months, we have some hopes to directly be able to utilize this amazing feature and skip hacky workarounds... 

Again, thanks a lot for all your efforts and great work!!

**Comment by mprahl** [comment]
> Hey everyone!
> 
> Thanks for the amazing work [@mprahl](https://github.com/mprahl) and the maintainers supporting you with your PRs! We were missing this feature so much and are currently looking into setting up Kubeflow and this is a major pain point for us.
> 
> Without intending any pressure and without any expectations of correctness: Might it be possible to get a guesstimate when this feature might be complete in MLFlow? And [@mprahl](https://github.com/mprahl): Do you have an idea when it could then be available / fully utilized in Kubeflow? Maybe I missed it in the thread, but is there a related Kubeflow ticket?
> 
> The reason I'm asking is that we are currently in the process of setting up Kubeflow with MLFlow and based on the quick progress on this over the last months, we have some hopes to directly be able to utilize this amazing feature and skip hacky workarounds...
> 
> Again, thanks a lot for all your efforts and great work!!

@andreas-repp I'm glad to hear you'll find this feature helpful! I can't give you a timeline, but I can give you the remaining items left to give you a sense. Here are the remaining to do's on the MLflow side:

1. Merge open PRs in the `orgnization-support` feature branch:
   1. #19571
   1. #19848
   1. #19778
1. Rebase the feature branch on `master`
1. Do extensive testing of the backend and SDK changes (big focus on backwards compatibility)
1. Merge the feature branch onto `master`
1. Then merge the UI enhancement #19468
1. Then handle additional UI improvements after a UXD review (particularly a workspace selection landing page).
1. Provide some admin tooling and/or documentation to allow reverting back a workspace enabled instance to non-workspace enabled.

On the Kubeflow side, I'll contribute an MLflow plugin that maps workspaces to Kubernetes namespaces, and maps all MLflow API endpoints to Kubernetes namespace scoped authorization. I have this working already, but am waiting to contribute it to Kubeflow until after the workspace feature is released.

**Comment by juliusvonkohout** [comment]
@mprahl do you separate between Kubeflow default-viewer and default-editor permissions for MLFlow?

**Comment by mprahl** [comment]
> [@mprahl](https://github.com/mprahl) do you separate between Kubeflow default-viewer and default-editor permissions for MLFlow?

The Kubernetes authorization plugin for MLflow will distinguish between view and edit permissions per resource type at the namespace level. My current implementation does not distinguish at the `resourceName` level to avoid additional database queries (e.g. only grant access to this experiment), but I'm reconsidering that.

**Comment by vitas** [comment]
Hi all,

I’ve been following the discussions around multi-tenancy, RBAC, and governance
for self-hosted MLflow.

As an experiment, I built an early open-source **Policy Enforcement Gateway**
that adds:

- tenant isolation for runs and model registry  
- minimal RBAC (viewer / contributor / admin)  
- OIDC / IAM integration  
- structured audit logging  

without modifying or forking MLflow itself.

Repository:
https://github.com/vitas/mlflow-enterprise-gateway

This is still an early implementation, and I’d really appreciate feedback
from teams operating MLflow in multi-team or regulated environments —
especially regarding API coverage and governance expectations.

Thanks for the discussion and the great work on MLflow.

**Comment by B-Step62** [comment]
MLflow 3.9 shipped [Workspace](https://mlflow.org/docs/latest/self-hosting/workspaces/getting-started/) feature that allow teams to share the same tracking server while logical boundary for experiments, prompts, models, etc. Please try it out and let us hear your feedbacks! 🚀
